cli/azd/internal/telemetry/storage.go

```diff
@@ -0,0 +1,277 @@
+package telemetry
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/benbjohnson/clock"
+)
+
+// A time.Format layout suitable for use in file names.
+const fsTimeLayout = "20060102T150405"
+
+// Max time before temp files are cleaned up
+const tempFileTtl = time.Duration(5) * time.Minute
+
+type Queue interface {
+	Enqueue(message []byte) error
+	EnqueueWithDelay(message []byte, delayDuration time.Duration, retryCount int)
+	Peek() (*StoredItem, error)
+	Remove(item *StoredItem) error
+}
+
+// StorageQueue is a FIFO-based queue backed by disk storage, with items stored as individual files.
+// The current implementation allows for multiple producers, single consumer.
+//
+// Items can be queued by Enqueue or EnqueueWithDelay.
+// QueueWithDelay allows for producers to queue items that should not be picked up
+// by consumers until after the specified duration has passed. This is useful for retry delay scheduling.
+//
+// Items can be read by Peek, which will read the next available item.
+// Once the item is processed, consumers are responsible for calling Remove to remove the item from the queue.
+type StorageQueue struct {
+	folder            string
+	itemFileExtension string
+
+	// Standard time library clock, unless mocked in tests
+	clock clock.Clock
+}
+
+type StoredItem struct {
+	// Number of retries attempted
+	retryCount int
+
+	// Message in the item
+	message []byte
+
+	// File name of the stored item
+	fileName string
+}
+
+func (itm *StoredItem) RetryCount() int {
+	return itm.retryCount
+}
+
+func (itm *StoredItem) Message() []byte {
+	return itm.message
+}
+
+type itemEntry struct {
+	name        string
+	readyTime   time.Time
+	fileModTime time.Time
+	retryCount  int
+}
+
+// Creates the storage-based queue.
+func NewStorageQueue(folder string, itemFileExtension string) (*StorageQueue, error) {
+	if err := os.MkdirAll(folder, osutil.PermissionDirectory); err != nil {
+		return nil, fmt.Errorf("failed to create storage queue folder: %v", err)
+	}
+
+	if !strings.HasPrefix(itemFileExtension, ".") {
+		itemFileExtension = "." + itemFileExtension
+	}
+
+	storage := StorageQueue{
+		folder:            folder,
+		itemFileExtension: itemFileExtension,
+		clock:             clock.New(),
+	}
+	return &storage, nil
+}
+
+// Queues a message.
+func (stg *StorageQueue) Enqueue(message []byte) error {
+	return stg.save(time.Duration(0), 0, message)
+}
+
+// Queues a message with delay.
+func (stg *StorageQueue) EnqueueWithDelay(message []byte, delayDuration time.Duration, retryCount int) error {
```

Comment:
So it looks like the client of the queue is supposed to maintain/update the retry count. Have you considered a design where the queue itself would advance the retry count and ensure that a message is re-tried at most N times?

Reply:
More specifically, how about the following
- `Enqueue()` operations include max allowable retries for the item
- Each `Peek()` operation increases the retry count automatically before the item is given to the queue client. When the count reaches maximum, the item is considered un-processable and discarded
- In addition to `Remove()` operation on an item, there is a `Delay()` operation that makes the item invisible to `Peek()` for given amount of time

Reply:
Increasing the retry count in the queue itself definitely increases robustness: max retry attempt is guaranteed by design instead of by contract (client-code honoring the design). I think the downsides would be: complexity in implementation (stored items are mutable for `Delay`), implicitness in message dropping might be unintuitive (example below).

I'll explain a little on how I think implicitness could be unintuitive. Consider the primary consumer of this queue being the telemetry client. The telemetry client stores a batch of telemetry items in each message and a subset of telemetry items can be retried, so partial list of items can be retried.

In the current design:
```golang
item := storage.Peek()
defer storage.Remove(item)
res := Handle(item.Message)
if !res.Success() {
    if res.CanRetry() {
        if res.IsPartialFailure() {
            eventsToRetry := res.GetFailedTelemetryEvents()
            storage.EnqueueWithDelay(eventsToRetry.Serialize(), res.RetryBackoff, 0)
        } else if res.IsCompleteFailure() {
            if item.RetryCount < maxRetryCount {
                storage.EnqueueWithDelay(item.Message(), time.Duration(0), item.RetryCount + 1)
            }
       }
    } else {
        // Log failure
    }
} else {
    // Handle success
}
```

With the queue enforcing the retry count:
```golang
item := storage.Peek()
shouldRemove := true
defer func() { if shouldRemove { storage.Remove(item) } }
res := Handle(item.Message)
if !res.Success() {
    if res.CanRetry() {
        if res.IsPartialFailure() {
            eventsToRetry := res.GetFailedTelemetryEvents()
            storage.EnqueueWithDelay(eventsToRetry.Serialize(), res.RetryBackoff, maxRetryCount)
        } else if res.IsCompleteFailure() {
            storage.Delay(item)
            shouldRemove = false
        }
    } else {
        // Log terminal failure
    }
} else {
    // Handle success
}
```

In the example with queue enforcing the retry count:
1. It's less clear when messages are dropped. We could also add immediate cleanup, where we check for max retry attempts and clean up immediately, but then we have repeated code.
2. We end up with needing to handle the mutable case of `Delay` separately than partial failures.

Overall, I don't think the robustness tradeoff is worth it given the extra dev time, less clear usability advantage and would lean towards our current simpler implementation. Let me know your thoughts.

Reply:
Given our usage pattern, I agree.